nixos/modules/services/misc/matrix-dendrite.nix

```diff
@@ -0,0 +1,171 @@
+    };
+    generatePrivateKey = lib.mkOption {
+      type = lib.types.bool;
+      default = true;
+      description = ''
+        Whether to generate a signing key.
+      '';
+    };
+    generateTls = lib.mkOption {
+      type = lib.types.bool;
+      default = false;
+      description = ''
+        Whether to generate a TLS certificate and key.
+      '';
+    };
+    tlsCert = lib.mkOption {
+      type = lib.types.nullOr lib.types.path;
+      example = "/var/lib/matrix-dendrite/server.cert";
+      default = null;
+      description = ''
+        The path to the TLS certificate. Will use pre-existing cert
+        if cfg.generateTls is not set true.
+      '';
+    };
+    tlsKey = lib.mkOption {
+      type = lib.types.nullOr lib.types.path;
+      example = "/var/lib/matrix-dendrite/server.key";
+      default = null;
+      description = ''
+        The path to the TLS key. Will use pre-existing key
+        if cfg.generateTls is not set true.
+      '';
+    };
+    settings = lib.mkOption {
+      type = lib.types.submodule {
+        freeformType = settingsFormat.type;
+        options.global = {
+          server_name = lib.mkOption {
+            type = lib.types.str;
+            example = "example.com";
+            description = ''
+              The domain name of the server, with optional explicit port.
+              This is used by remote servers to connect to this server.
+              This is also the last part of your UserID.
+            '';
+          };
+          private_key = lib.mkOption {
+            type = lib.types.nullOr lib.types.path;
+            example = "${workingDir}/matrix_key.pem";
+            default = "${workingDir}/matrix_key.pem";
+            description = ''
+              The path to the signing private key file, used to sign
+              requests and events.
+            '';
+          };
+          trusted_third_party_id_servers = lib.mkOption {
+            type = lib.types.listOf lib.types.str;
+            example = [ "matrix.org" ];
+            default = [ "matrix.org" "vector.im" ];
+            description = ''
+              Lists of domains that the server will trust as identity
+              servers to verify third party identifiers such as phone
+              numbers and email addresses
+            '';
+          };
+          disable_federation = lib.mkOption {
+            type = lib.types.bool;
+            default = true;
+            description = ''
+              Disables federation. Dendrite will not be able to make
+              any outbound HTTP requeststo other servers and the federation
+              API will not be exposed.
+            '';
+          };
+        };
+        options.client_api = {
+          registration_disabled = lib.mkOption {
+            type = lib.types.bool;
+            default = true;
+            description = ''
+              Whether to disable user registration to the server
+              without the shared secret.
```

Comment:
If setting such a shared secret is a relatively common operation, I suggest adding another option (outside of `settings`) for securely specifying `client_api.registration_shared_secret` with a file. E.g.:
```nix
{
  options.services.matrix-dendrite.registrationSecretFile = mkOption {
    type = types.nullOr types.path;
  };

  config = mkIf (cfg.registrationSecretFile != null) {
    services.matrix-dendrite.settings.client_api.registration_shared_secret = "NIXOS_REGISTRATION_SECRET";
 
    systemd.services.matrix-dendrite.script = ''
      registrationSecret=$(cat ${cfg.registrationSecretFile})
      configFile=$(mktemp)
      sed "s/NIXOS_REGISTRATION_SECRET/$registrationSecret/" ${configurationYaml} > "$configFile"
      dendrite-monolith-server --config "$configFile"
    '';
  };
}
```

Reply:
I made the changes, but right now it will fail if cfg.registrationSecretFile is null

Reply:
Think I fixed it, pushing soon

Reply:
Why does it fail?

Reply:
cfg.registrationSecretFile is set to null by default, so the string substitution fails

Reply:
I just pushed an option check, let me know what you think

Reply:
Ah well you'll have to use a different start script depending on if it's set to `null`

Reply:
Note the `mkIf (cfg.registrationSecretFile != null)` in my example

Reply:
Something like
```nix
{
  script = ''
    ${if (cfg.registrationSecretFile == null) ''
      configFile=$(mktemp)
      ...
    '' else
      configFile=${configurationYaml}
    ''}
    dendrite-monolith-server ... 
  '';
}
```

Reply:
Hmm.. I'm afraid that the `sed` invocation could have escaping problems for special characters in the secret

Reply:
I was worried about that too, what do you think though of the optional check? I think it works? 

Re: sed, I'll look into seeing if there's a better option. Maybe jq? 

```
echo '{"hello": "world"}' | jq --arg foo bar '. + {foo: $foo}'
# {
#   "hello": "world",
#   "foo": "bar"
# }
```

Reply:
Regarding `sed`, this is a bit obtuse, but maybe something like a `sed` to change `NIXOS_REGISTRATION_SECRET` to `@nixos_registration_secret@` and then a 
```
(substituteAll {
  src = "${configurationYaml}";
  nixos_registration_secret = "${registrationSecret}";
})
```
would do it? I only looked at this for about two minutes so I expect this to be a poor solution, but thought you might benefit from being pointed at `substituteAll` in this case.

Reply:
@Infinisil I think the only issue I can find is with \, but generally those aren't used in passwords anyways. @Wulfsta I don't think that solves the current issue.

Reply:
I guess the only way to use that function would be to write up a `start-dendrite.nix` that overwrites the `shellHook` of `mkShell` and then:
```
...
settings.client_api.registration_shared_secret = "@nixos_registration_secret@";
...
script = ''
  nix-shell /path/to/start-dendrite.nix
''
...
```
Not sure that would be the best solution for this, but it was interesting to think about.

Reply:
How about using environmental variables and `envsubst`? That should handle most escaping issues, apart from maybe YAML's literals, which you could handle with `jq` since YAML is a superset of JSON.

I haven't tested this, but maybe something like this?

```
settings.client_api.registration_shared_secret = "$REGISTRATION_SHARED_SECRET";
```
```
REGISTRATION_SHARED_SECRET="$(cat "$secret_path" | jq -aR . )"
envsubst -i "$input"  -o "$output"
```